datafusion/src/optimizer/projection_push_down.rs

```diff
@@ -417,9 +416,15 @@ fn optimize_plan(
                 })
                 .collect::<Result<Vec<_>>>()?;
 
+            let mut new_schema = Vec::new();
+            for df_field in schema.fields() {
+                if union_required_fields.contains(df_field.field()) {
```

Comment:
This has quadratic complexity, which can be problematic when we have many fields but I think we have more areas in the code base with this issue, so not really needed to solve now.

Reply:
given that `union_required_fields` is a hashset, shouldn't the cost be close to constant?